doc/config_rsmexplain.rst.inc

```diff
@@ -0,0 +1,66 @@
+.. _config_file_rsmexplain:
+
+Experiment configuration file
+"""""""""""""""""""""""""""""
+
+This is a file in ``.json`` format that provides overall configuration options for an ``rsmexplain`` experiment. Here's an `example configuration file <https://github.com/EducationalTestingService/rsmtool/blob/main/examples/rsmexplain/config_rsmexplain.json>`__ for ``rsmexplain``.
+
+.. note:: To make it easy to get started with  ``rsmexplain``, we provide a way to **automatically generate** configurations file both interactively as well as non-interactively. Novice users will find interactive generation more helpful while more advanced users will prefer non-interactive generation. See :ref:`this page <autogen_configuration>` for more details.
+
+Next, we describe all of the ``rsmexplain`` configuration fields in detail. There are four required fields and the rest are all optional. We first describe the required fields and then the optional ones (sorted alphabetically).
+
+background_data
+~~~~~~~~~~~~~~~
+The path to the background data feature file one of the :ref:`supported formats <input_file_format>`. Each row should correspond to a single response and contain the numeric feature values extracted for this response. In addition, there should be a column containing a unique identifier (ID) for each response. This path can be absolute or relative to the location of config file. It must contain at least 300 responses to ensure meaningful explanations.
+
+explainable_data
+~~~~~~~~~~~~~~~~
+The path to the file containing the data that we want to explain. The file should be in one of the :ref:`supported formats <input_file_format>`. Each row should correspond to a single response and contain numeric feature values extracted for this response. In addition, there should be a column containing a unique identifier (ID) for each response. The path can be absolute or relative to the location of config file.
+
+experiment_id
+~~~~~~~~~~~~~
+An identifier for the ``rsmexplain`` experiment. This will be used to name the report. It can be any combination of alphanumeric values, must *not* contain spaces, and must *not* be any longer than 200 characters.
+
+experiment_dir
+~~~~~~~~~~~~~~
+The directory containing the rsmtool models we want to explain. This directory should contain an ``output`` sub-directory and that sub-directory should contain two files: the ``<experiment_id>.model`` and  ``<experiment_id>_feature.csv``. Note that ``<experiment_id>`` refers to the one defined in this same configuration file. As an example of this directory structure, take a look at the ``existing_experiment`` directory  `here <https://github.com/EducationalTestingService/rsmtool/tree/main/tests/data/experiments/svr-explain>`__
+
+background_kmeans_size *(Optional)*
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The size of the k-means sample for background. Defaults to ``500``. A dataset is summarized with k mean samples weighted by the number of data points they each represent.
```

Comment:
Is there a rule-of-thumb for setting this number based on data characteristics?

Reply:
@remo-help I think you did a bunch of experiments to come up with this number or is this recommended by SHAP? Adding some details as to how we came up with this number would be useful, I agree.

Reply:
This is very helpful! Thanks @remo-help. @damien2012eng let's add this information.

Reply:
Thanks @remo-help! 
Is there a threshold for the sample size above which the running time is significantly slow?
Also can we recommend not to use sampling when the background data size is lower than some threshold (500 like the current default or maybe more)?

Reply:
Thanks for the detailed explanation @remo-help!

@damien2012eng @desilinguist we say here that it's an optional value but we use it with no default here: https://github.com/EducationalTestingService/rsmtool/blob/main/rsmtool/rsmexplain.py#L307

Also we cluster on the masked dataset, meaning if the user specified sample size or range,  the input for kmeans will be on that sampled data.

Reply:
Good catch on the optional. 

I think that's how @remo-help had implemented it, IIRC. @remo-help please let us know if masking the featureset _before_ doing k-means clustering is the wrong thing to do?

Reply:
Actually, we are calling `mask()` without the `feature_range` parameter so that means all of the features are selected. We still call `mask()` since it also filters out feature columns that didn't end up being used by the model.

Reply:
right! yes that seems correct.

Reply:
@desilinguist  that's correct, `mask()` also makes sure the array is of the right type.
@tamarl08 in the line you reference, I'm pretty sure the default is specified in the config file, so if the user doesnt specify anything, the value pulled here will automatically be the default

Reply:
yep, @remo-help  is right. The default value is automatically set when the configuration is parsed and nothing is specified by the user.

Reply:
can we set the default to min(background size, 500)?
I can't find the correct `shap.kmeans` method, it may be ok to pass the a `k` larger than `n`.

Reply:
Here's the method: https://github.com/slundberg/shap/blob/master/shap/utils/_legacy.py#L10

It's calling `KMeans` from scikit-learn. I expect it would raise an error if this were the case. @damien2012eng  can we test this by setting the kmeans parameter to a value larger than the background data size?

Reply:
`shap.kmeans()` is inconveniently not present in the SHAP documentation, although they use it in the examples.
I wouldn't be surprised if it was a simple wrapper around the `sklearn.cluster.KMeans` function: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html#sklearn.cluster.KMeans

Reply:
@desilinguist  is right. Here is the raised error from the sklearn when I set the kmeans_size to be `750`. 

> ValueError: n_samples=500 should be >= n_clusters=750.